RSS for shelves

mattkatz opened the pull request:
This seems to correctly implement an RSS feed for shelves, closing  #2871,

jaschaurbach commented:
@mouse-reeve any idea why the test fails?

mattkatz commented:
@jaschaurbach good news is that it looks to be a very similar error from the failed test on github vs what I saw locally. Somehow I'm not mocking something out well enough and the test is trying to publish things to redis etc - none of that is set up in the test environment so everything goes boom. I hope to revisit this sometime when I can spend time figuring out why my test explodes. 

It's frustrating to get the logic working locally but not the test, but for a project like this where it's just... some people ... working on it, I sympathize that there isn't time to triage the work of everyone that drops by with a pull request. If I can fix it, I will. If mouse or someone else competent sees what I'm doing wrong, that's great.

If I just need to pull an RSS feed of everything for my purposes and then filter out to just the to-read items, I can do that on my side.

hughrun commented on bookwyrm/tests/views/test_rss_feed.py:
```diff
@@ -127,3 +127,26 @@ def test_rss_quotation_only(self, *_):
+
+    def test_rss_shelf(self, *_):
+        """load the rss feed of a shelf"""
+        with patch("bookwyrm.models.activitypub_mixin.broadcast_task.apply_async"):
```
You need to add another mock here for `bookwyrm.activitystreams.add_book_statuses_task`:
```python
with patch("bookwyrm.models.activitypub_mixin.broadcast_task.apply_async"), patch(
"bookwyrm.activitystreams.add_book_statuses_task.delay"
)
```

mattkatz commented:
@hughrun - all checks passed - hopefully the commits are not controversial.

hughrun commented:
@mattkatz the title of this is still 'DRAFT' - is this ready for review or are you still working on it?

mattkatz commented:
> @mattkatz the title of this is still 'DRAFT' - is this ready for review or are you still working on it?

Fixed. This is ready for review.

hughrun commented:
@bookwyrm-social/code-review if someone has time to review this that would be awesome 🙂

MaggieFero commented:
This sounds so useful! I hope somebody has a chance to review it soon.

dato commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,66 @@ def item_link(self, item):
+
+    def items(self, obj):
+        """the user's activity feed"""
+        return obj.books.order_by("-published_date")[:10]
```
Should sort by shelved date, right?:

```suggestion
        return obj.books.order_by("-shelfbook__shelved_date")[:10]
```

dato reviewed (commented):
Hello! Thank you for this feature, and apologies the review took so long. I'm sending you my suggestions, minus i18n issues (which I'll do in a second pass).

Most of these suggestions are minor, or self-explanatory. Please let me know what you think.

mattkatz commented:
@dato - I'll take a look. And thank YOU for volunteering time to do this. I know this is real work and effort and I appreciate it. I'm checking it out now.

mattkatz commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,66 @@ def item_link(self, item):
+            desc = f"  {obj.description}"
+        return f"Books added to the '{obj.name}' shelf for {obj.user.name}.{desc}"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        # TODO: gotta check that this is the SHELF user!
+        context["user"] = kwargs["request"].user
+        context["shelf"] = kwargs["obj"]
+        return context
```
I think this was an artifact of how things were done in November, if I had to guess. I was cribbing from somewhere, I promise I'm not having real original thoughts.
But as of now...
```
mk@miter [10:01:18 PM] [~/projects/bookwyrm] [rss-for-shelves *]
-> % pwd
/home/mk/projects/bookwyrm
mk@miter [10:01:39 PM] [~/projects/bookwyrm] [rss-for-shelves *]
-> % ag get_context_data
bookwyrm/views/rss_feed.py
238:    def get_context_data(self, **kwargs):
239:        context = super().get_context_data(**kwargs)
mk@miter [10:02:53 PM] [~/projects/bookwyrm] [rss-for-shelves *]
-> %

```
So I'm deleting it.

mattkatz commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,66 @@ def item_link(self, item):
+        # Shelf.privacy_filter(request.user).filter(user=user).all()
+
+        # TODO: get the SHELF of the object
+        shelf = get_object_or_404(user.shelf_set, identifier=shelf_identifier)
```
resolved with a commit plus updated comments to match.

mattkatz commented:
@dato thanks for the thorough review. I've addressed all comments. 
I haven't added new tests around ordering  - I don't know that it's worth the test load. If it's needed I'd update the test rss to add 2 books, figure out how to add a published date to each, then add the them to the shelf and expect that ordering is maintained.

dato commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,61 @@ def item_link(self, item):
+
+    def title(self, obj):
+        """title of the rss feed entry"""
+        return _(f"Books added to {obj.name}")
```
Since this is the title RSS readers will show, how about including the shelf owner in a concise manner:

```suggestion
        return _(f"{obj.user.name}’s {obj.name} shelf")
```

dato commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,61 @@ def item_link(self, item):
+    def description(self, obj):
+        """description of the shelf including the shelf name and user."""
+        # if there's a description, lets add it. Not everyone puts a description in.
+        desc = ""
+        if obj.description:
+            desc = f"  {obj.description}"
+        return f"Books added to the '{obj.name}' shelf for {obj.user.name}.{desc}"
```
Similarly here (plus adding missing call to `_()`):

```suggestion
        if desc := obj.description:
            return _(f"{obj.user.name}’s {obj.name} shelf: {desc}"
        return _(f"Books added to {obj.user.name}’s {obj.name} shelf"
```

dato commented on bookwyrm/templates/rss/edition.html:
```diff
@@ -0,0 +1,4 @@
+{{obj.title}} by {{obj.author_text}}
```
I believe this should be wrapped with `{% blocktrans %} … {% endblocktrans %}`.

I still want to improve/review this format a bit, but I say we merge the PR with the two suggestions above, and do a follow-up (unless you prefer one extra round here.

Basically what I want to review is the ISBN labels: not sure if we want to show both if both are available?; and check whether newlines are needed somewhere, for readability in RSS readers.)

mattkatz replied:
I can add the wrapping no problem. I would like to merge soon so I can see some incremental value. The i8n effort I think was merged after I started this PR and I want to merge before folks start any new initiatives!

mattkatz replied:
wrapping is done in 1d8bd2be892dde3079915a6397e28edeba1c545f

mattkatz replied:
in forms we have multiple places to put the isbn, both 10 and 13 - but in the templates for books and editions I find that we handle book identifiers in https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/templates/book/book_identifiers.html 
In that, we ONLY display ISBN_13. I'm fine to amend this to show each identifier as it exists similar to the book_identifiers helper

mattkatz replied:
amended to have no ISBN10  but the other identifiers in book_identifiers in 09d857e6fbf731283d4568f1ac4354e0120722ca

dato replied:
this is excellent, thank you <3

dato reviewed (approved):
Looks good to me with the first two suggestions below.

dato commented:
Ah one last thing for a follow-up PR: producing a `<link rel="alternate" type="application/rss+xml" href="…" title="…" />` element in the template for public shelves (unlisted too?).

dato commented on bookwyrm/templates/rss/edition.html:
```diff
@@ -0,0 +1,8 @@
+{% blocktrans %}{{{obj.title}} by {{obj.author_text}}{% endblocktrans %}
```
My bad, I didn't quite understand how _blocktrans_ works:

```suggestion
{% load i18n %}
{% blocktrans trimmed with book_title=obj.title book_author=obj.author_text %}
  ‘{{ book_title }}’ by {{ book_author }}
{% endblocktrans %}
```

dato commented on bookwyrm/templates/rss/edition.html:
```diff
@@ -0,0 +1,8 @@
+{% blocktrans %}{{{obj.title}} by {{obj.author_text}}{% endblocktrans %}
+{{obj.description|default:""}}
+{% if obj.description %}{% trans "ISBN13:" %} {{item.isbn_13|default: ""}}{% endif %}
+{% if obj.description %}{% trans "OCLC Number:" %} {{item.oclc_number|default: ""}}{% endif %}
+{% if obj.description %}{% trans "ASIN:" %} {{item.asin|default: ""}}{% endif %}
+{% if obj.description %}{% trans "Audible ASIN:" %} {{item.aasin|default: ""}}{% endif %}
+{% if obj.description %}{% trans "ISFDB ID:" %} {{item.isfdb|default: ""}}{% endif %}
+{% if obj.description %}{% trans "Goodreads:" %} {{item.goodreads_key|default: ""}}{% endif %}
```
`|default: ""`  must become `|default:""` (no spacing). However, how about having the conditionals apply to each metadata item?:


```suggestion
{% if obj.isbn_13 %}{% trans "ISBN 13:" %} {{ obj.isbn_13 }}{% endif %}
{% if obj.oclc_number %}{% trans "OCLC Number:" %} {{ obj.oclc_number }}{% endif %}
{% if obj.asin %}{% trans "ASIN:" %} {{ obj.asin }}{% endif %}
{% if obj.aasin %}{% trans "Audible ASIN:" %} {{ obj.aasin }}{% endif %}
{% if obj.isfdb %}{% trans "ISFDB ID:" %} {{ obj.isfdb }}{% endif %}
{% if obj.goodreads_key %}{% trans "Goodreads:" %} {{ obj.goodreads_key }}{% endif %}
```

dato reviewed (approved):
Hi again—

> I would like to merge soon so I can see some incremental value

Yes, I support this as well.

I think we can merge, with the tiny adjustments below (needed for a green build).

hughrun commented:
I've spotted a few issues, review coming shortly.

hughrun commented on bookwyrm/templates/rss/edition.html:
```diff
@@ -0,0 +1,11 @@
+{% load i18n %}
+{% blocktrans trimmed with book_title=obj.title book_author=obj.author_text %}
+  ‘{{ book_title }}’ by {{ book_author }}
```
Need to cover cases where there is no author listed. Something like:

`{{ book_title }}’{% if book_author %}by {{ book_author }}{% endif %}`

mattkatz replied:
handled in  98724dfa4774e2103e3a80b6b835959bef422ddaea68dce98b

hughrun commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,60 @@ def item_link(self, item):
+
+    def title(self, obj):
+        """title of the rss feed entry"""
+        return _(f"{obj.user.name}’s {obj.name} shelf")
```
This should be `obj.user.display_name` otherwise it can render as `None`.

mattkatz replied:
done in  98724dfa47

hughrun commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,60 @@ def item_link(self, item):
+        """description of the shelf including the shelf name and user."""
+        # if there's a description, lets add it. Not everyone puts a description in.
+        if desc := obj.description:
+            return _(f"{obj.user.name}’s {obj.name} shelf: {desc}")
```
This should be `obj.user.display_name` as well.

mattkatz replied:
done in  98724dfa47

hughrun commented on bookwyrm/templates/rss/edition.html:
```diff
@@ -0,0 +1,11 @@
+{% blocktrans trimmed with book_title=obj.title book_author=obj.author_text %}
+  ‘{{ book_title }}’ by {{ book_author }}
+{% endblocktrans %}
+{{obj.description|default:""}}
```
This shouldn't be a blocker because it seems to be a bit difficult to manage embedded HTML in the Django syndication framework, but I noticed this just runs on without line breaks so it renders like `Title of Book by Jo Bloggs Here is a description of the book`.

mattkatz replied:
I'm not sure what to do here. 

If I could figure out how to trigger a cdata block, we could do something to embed new lines. 
This seems like an approach someone got working, if it makes sense to you, I can take a swing at it.
https://nemecek.be/blog/33/how-to-create-rss-feed-with-html-content-in-django

For my purposes, I'm very interested in getting the book ISBN and other identifiers in as raw a format as possible, but I'd like to make sure this feed is useful and looks good for humans in a feed reader.

When I look at how other book tracking sites handle this, they use the cdata block pretty liberally. 
They also just introduce custom elements like isbn and asin directly in the feed. Which doesn't seem valid xml but it seems not to blow anything up.
I'm not ready to dedicate myself just yet to implementing a custom feed generator as described in https://docs.djangoproject.com/en/5.0/ref/contrib/syndication/#custom-feed-generators

mattkatz replied:
btw @hughrun - this is listed as a blocking change I think - assuming I'm not misreading githubs UI.

hughrun replied:
I'll take a look at this shortly.

hughrun reviewed (changes requested):
This generally seems to work, just a couple of things to clean up.

mattkatz commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,60 @@ def item_link(self, item):
+        if hasattr(item, "pure_name") and item.pure_name:
+            return item.pure_name
+        authors = item.authors
+        authors.display_name = f"{item.author_text}:"
```
handled in a0d15ccec0

mattkatz commented on bookwyrm/views/rss_feed.py:
```diff
@@ -177,3 +178,60 @@ def item_link(self, item):
+
+    def item_title(self, item):
+        """render the item title"""
+        if hasattr(item, "pure_name") and item.pure_name:
```
handled in 8dc412c4cbbb3596b1bb6c59d11e6c191050cbe4

hughrun commented:
@mattkatz I've made a [PR in your repository](https://github.com/mattkatz/bookwyrm/pull/1) with a fix for the template issue, and added a link the head to enable auto-discovery of the feeds.

Sorry, I kind of lead you in the wrong direction a bit with adding logic for missing authors. We can't put logic inside block tags. But when I thought about it more I realised we shouldn't be translating titles and descriptions anyway: e.g. if it's a book in French the title should be in French, not in whatever language I have my device set to display.

If you're happy with my changes just merge them in to your branch and they will flow through to this PR.

hughrun commented:
@bookwyrm-social/code-review can someone please double-check this, since I added the last commit here?

hughrun commented:
Sorry this took so long @mattkatz

mattkatz commented:
Thanks for the help @hughrun - nobody owes me anything, thank you for all the time you spent and spend!
